File: 300-longest-increasing-subsequence/300-longest-increasing-subsequence.py
class Solution:
  def lengthOfLIS(self, nums: List[int]) -> int:
    buffer = []
    for n in nums:
      l = 0
      r = len(buffer)
      while l < r:
        m = (l + r) // 2
        if buffer[m] < n: l = m + 1
        else: r = m
      if l >= len(buffer):
        buffer.append(n)
      else:
        buffer[l] = n
    return len(buffer)

File: 300-longest-increasing-subsequence/300-longest-increasing-subsequence.cpp
class Solution {
public:
  int lengthOfLIS(vector<int>& nums) {
    vector<int> dp(nums.size());
    int ans = 1;
    dp[0] = 1;
    for (int i = 1; i < nums.size(); i++) {
      int v = nums[i];
      int res = 1;
      for (int j = 0; j < i; j++) {
        if (v > nums[j]) {
          res = max(dp[j]+1, res);
        }
      }
      dp[i] = res;
      if (res > ans) {
        ans = res;
      }
    }
    return ans;
  }
};